### ISSUE TITLE
no-multi-assign option ignore some operators.

### ISSUE THREAD

**Issue by magic-akari** [eslint/eslint#14348]
**What rule do you want to change?**

no-multi-assign
https://eslint.org/docs/rules/no-multi-assign

**Does this change cause the rule to produce more or fewer warnings?**

For this rule, it produces fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**

I think it should be an option, something like `ignoreOperators`  of type `Array<string>`. When set with some values, no-multi-assign will only trigger when operators is not in the ignore list.

**Please provide some example code that this change will affect:**

```typescript
interface IList<T = unknown> {
  list: T[];
  hasMore: boolean;
}

export const emptyList = <T>(): IList<T> => {
  return {
    list: [],
    hasMore: true,
  };
};

export const addItem = <T>(listMap: Record<string, IList<T> | undefined>, key: string, item: T) => {
  const result = (listMap[key] ??= emptyList<T>());
  //                            ^ Unexpected chained assignment.
  result.list.push(item);
};
```

```typescript
export const addItem = <T>(listMap: Record<string, IList<T> | undefined>, key: string, item: T) => {
  listMap[key] ??= emptyList<T>();

  const result = listMap[key];
  result.list.push(item);
  // ^ Object is possibly 'undefined'.ts(2532)
};
```

```typescript
export const addItem = <T>(listMap: Record<string, IList<T> | undefined>, key: string, item: T) => {
  listMap[key] ??= emptyList<T>();

  const result = listMap[key]!;
  //                         ^ Forbidden non-null assertion.eslint@typescript-eslint/no-non-null-assertion
  result.list.push(item);
};
```

```typescript
export const addItem = <T>(listMap: Record<string, IList<T> | undefined>, key: string, item: T) => {
  listMap[key] ??= emptyList<T>();

  const result = listMap[key] ?? emptyList<T>(); // everything is ok, but it is a bit noisy.
  result.list.push(item);
};
```

**What does the rule currently do for this code?**

see the example above.

**What will the rule do after it's changed?**

You will have the option to ignore errors with specific operators such as `??=` or `||=`.

**Are you willing to submit a pull request to implement this change?**

Yes, I think is is a simple change. I will create a PR later.

**Comment by nzakas** [comment]
I’m sorry, it’s not clear what you are proposing. Please explain more about what you’d like to see changed. What is the use case? What can’t you do right now?

Also, can you please provide some examples for regular JavaScript code rather than TypeScript?

**Comment by magic-akari** [comment]
Sorry, everything is OK in Javascript. But in TypeScript, I'm stuck with the type checking. So I can't provide Javascript code.

**Comment by nzakas** [comment]
ESLint doesn’t natively support TypeScript, so without a reproducible case in JavaScript we can’t look into this any further. If you are having trouble with TypeScript-specific issues, I’d suggest opening an issue on the typescript-eslint project.
